crates/flux-infer/src/refine_tree.rs

```diff
@@ -353,31 +354,41 @@ impl std::ops::Deref for NodePtr {
 }
 
 impl Node {
-    fn simplify(&mut self, genv: GlobalEnv) {
-        for child in &self.children {
-            child.borrow_mut().simplify(genv);
-        }
-
+    fn simplify(&mut self, genv: GlobalEnv, mut assumed_preds: FxHashSet<Expr>) {
+        // First, simplify the node itself
         match &mut self.kind {
             NodeKind::Head(pred, tag) => {
-                let pred = pred.normalize(genv).simplify();
+                let pred = pred.normalize(genv).simplify(&assumed_preds);
                 if pred.is_trivially_true() {
                     self.kind = NodeKind::True;
                 } else {
                     self.kind = NodeKind::Head(pred, *tag);
                 }
             }
-            NodeKind::True => {}
             NodeKind::Assumption(pred) => {
-                *pred = pred.normalize(genv).simplify();
-                self.children
-                    .extract_if(.., |child| {
-                        matches!(child.borrow().kind, NodeKind::True)
-                            || matches!(&child.borrow().kind, NodeKind::Head(head, _) if head == pred)
-                    })
-                    .for_each(drop);
+                *pred = pred.normalize(genv).simplify(&assumed_preds);
+                pred.flatten_conjs().into_iter().for_each(|conjunct| {
+                    assumed_preds.insert(conjunct.erase_spans());
+                });
             }
-            NodeKind::Trace(_) | NodeKind::Root(_) | NodeKind::ForAll(..) => {
+            _ => {}
+        }
+
+        // Then simplify the children
+        // (the order matters here because we need to collect assumed preds first)
+        for child in &self.children {
+            // Need to clone the assumed_preds here because it may (and probably
+            // will) diverge across branches
+            child.borrow_mut().simplify(genv, assumed_preds.clone());
```

Comment:
You can implement this with a [`SnapshotMap`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/snapshot_map/struct.SnapshotMap.html) and avoid the `clone` and quadratic behavior.

Reply:
Haskell: "Look what they have to do to mimic a fraction of our power"